### PR TITLE
Fix ExprDifference issues

### PR DESCRIPTION
**Pull request by APickledWalrus**
### Description
This PR aims to fix some issues, mainly with the initialization stage of ExprDifference.

There were multiple issues, but the main one (which is why https://github.com/SkriptLang/Skript/pull/5331 had test failures) is that it does not properly compare *known* types. For example, if we had two types, `string` and `number`, the resolved class info is `object` (it's the shared superclass). Of course, this expression has handling for variables that allows `object` as a valid classinfo. However, this meant that the `string` and `number` difference would be accepted. This caused Skript to interpret the following line incorrectly in some cases:
`difference between "%{_a}%" parsed as a number and "%{_b}%" parsed as a number`
It should be interpreted as:
`difference between ("%{_a}%" parsed as a number) and ("%{_b}%" parsed as a number)`
However, because of the faulty checks, it would sometimes be interpreted as:
`(difference between ("%{_a}%" parsed as a number) and "%{_b}%") parsed as a number`

So, I've cleaned up the class, and the new logic is as follows:

If we have two known types:
- Attempt to convert them to each other - fail if no conversions can be made

If we have one known type:
- Attempt to convert object to the known type. I believe this will always succeed with new converters, but it also allows us to verify that the known type can actually have a difference operation performed on it

If we have no known types:
- Resolve during runtime when we know types

In the PR I mentioned above, ExprParse was sometimes registered before ExprDifference which is why this error occurred. It's not problematic right now because ExprDifference is registered before ExprParse (but we shouldn't depend on load order to prevent bugs!!)

**Thorough review of the logic and additional testing is likely needed**

---
**Target Minecraft Versions:** Any
**Requirements:** None
**Related Issues:** none?

### CODE REVIEW

**Review by kiip1 — APPROVED** [review]
Thanks :)

File: src/main/java/ch/njol/skript/expressions/ExprDifference.java [hunk 1]
```diff
@@ -33,123 +28,159 @@
...
-			relativeType = ci.getMathRelativeType();
+			math = classInfo.getMath();
+			relativeType = classInfo.getMathRelativeType();
 		}
+
+		this.first = first;
+		this.second = second;
+
+		if (fail) {
+			Skript.error("Can't get the difference of " + CondCompare.f(first) + " and " + CondCompare.f(second), ErrorQuality.SEMANTIC_ERROR);
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprDifference.java:138]: It's `SEMANTIC_ERROR` by default
```suggestion
			Skript.error("Can't get the difference of " + CondCompare.f(first) + " and " + CondCompare.f(second));
```

**Review by AyhamAl-Ali — APPROVED** [review]
Nice PR my associate ⚡

File: src/main/java/ch/njol/skript/expressions/ExprDifference.java [hunk 2]
```diff
@@ -33,123 +28,168 @@
...
+	"if difference between {command::%player%::lastuse} and now is smaller than a minute:",
+	"\tmessage \"You have to wait a minute before using this command again!\""
+})
 @Since("1.4")
 public class ExprDifference extends SimpleExpression<Object> {
 	
 	static {
-		Skript.registerExpression(ExprDifference.class, Object.class, ExpressionType.COMBINED, "difference (between|of) %object% and %object%");
+		Skript.registerExpression(ExprDifference.class, Object.class, ExpressionType.COMBINED,
+			"difference (between|of) %object% and %object%"
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprDifference.java:53]:
```suggestion
				"difference (between|of) %object% and %object%"
```

File: src/main/java/ch/njol/skript/expressions/ExprDifference.java [hunk 3]
```diff
@@ -33,123 +28,168 @@
...
-/**
- * @author Peter Güttinger
- */
 @Name("Difference")
 @Description("The difference between two values, e.g. <a href='./classes.html#number'>numbers</a>, <a href='./classes/#date'>dates</a> or <a href='./classes/#time'>times</a>.")
-@Examples({"if difference between {command::%player%::lastuse} and now is smaller than a minute:",
-		"\tmessage \"You have to wait a minute before using this command again!\""})
+@Examples({
+	"if difference between {command::%player%::lastuse} and now is smaller than a minute:",
+	"\tmessage \"You have to wait a minute before using this command again!\""
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprDifference.java:46]:
```suggestion
		"\tmessage \"You have to wait a minute before using this command again!\""
```

File: src/main/java/ch/njol/skript/expressions/ExprDifference.java [hunk 4]
```diff
@@ -33,123 +28,168 @@
...
-			ci = Classes.getSuperClassInfo(Utils.getSuperType(first.getReturnType(), second.getReturnType()));
-		}
-		assert ci != null;
-		if (!ci.getC().equals(Object.class) && ci.getMath() == null) {
-			Skript.error("Can't get the difference of " + CondCompare.f(first) + " and " + CondCompare.f(second), ErrorQuality.SEMANTIC_ERROR);
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		Expression<?> first = LiteralUtils.defendExpression(exprs[0]);
+		Expression<?> second = LiteralUtils.defendExpression(exprs[1]);
+		if (!LiteralUtils.canInitSafely(first, second)) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprDifference.java:71]: Doesn't canInitSafely go before defending?

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprDifference.java (hunk 4 above)
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprDifference.java:71]: No, IIRC canInitSafely is to make sure that defending them was actually successful

**Review by UnderscoreTud — APPROVED** [review]

**Review by sovdeeth — APPROVED** [review]

**Pull request merged**